Binary files u-boot.orig/.git/index and u-boot/.git/index differ
diff -rupbN u-boot.orig/1.patch u-boot/1.patch
--- u-boot.orig/1.patch	1970-01-01 00:00:00.000000000 +0000
+++ u-boot/1.patch	2025-07-08 16:47:19.895985940 +0000
@@ -0,0 +1,109 @@
+From fac92a6b8006d2678a3eb1cae895f8d6471739b3 Mon Sep 17 00:00:00 2001
+From: Hugh Cole-Baker <sigmaris@gmail.com>
+Date: Mon, 19 Aug 2024 16:30:56 +0000
+Subject: [PATCH] WIP qcom configs
+
+---
+ board/qualcomm/default.env | 11 +++++++----
+ configs/qcom_defconfig     | 25 ++++++++++++++++++++-----
+ 2 files changed, 27 insertions(+), 9 deletions(-)
+
+diff --git a/board/qualcomm/default.env b/board/qualcomm/default.env
+index dbf6f4e72600..9647f68a73dd 100644
+--- a/board/qualcomm/default.env
++++ b/board/qualcomm/default.env
+@@ -3,9 +3,12 @@ stdout=serial,vidconsole
+ stderr=serial,vidconsole
+ preboot=scsi scan; usb start
+ fastboot=fastboot -l $fastboot_addr_r usb 0
+-do_boot=bootefi bootmgr
++do_boot=bootflow scan -b
++serial_gadget=setenv stdin serial,button-kbd,usbacm; setenv stdout serial,vidconsole,usbacm; setenv stderr serial,vidconsole,usbacm; setenv bootretry -1; echo Enabled U-Boot console gadget :3
+ bootmenu_0=Boot first available device=run do_boot
+-bootmenu_1=Enable fastboot mode=run fastboot
+-bootmenu_2=Reset device=reset
+-menucmd=bootmenu
++bootmenu_1=Enable USB mass storage=ums 0 mmc 0
++bootmenu_2=Enable fastboot mode=run fastboot
++bootmenu_3=Enable serial console gadget=run serial_gadget
++bootmenu_4=Reset device=reset
++menucmd=bootmenu -1
+ bootcmd=run do_boot
+diff --git a/configs/qcom_defconfig b/configs/qcom_defconfig
+index 9cf450028fe9..1c2468ff7cf0 100644
+--- a/configs/qcom_defconfig
++++ b/configs/qcom_defconfig
+@@ -4,13 +4,15 @@ CONFIG_POSITION_INDEPENDENT=y
+ CONFIG_SYS_INIT_SP_BSS_OFFSET=1572864
+ CONFIG_ARCH_SNAPDRAGON=y
+ CONFIG_NR_DRAM_BANKS=24
+-CONFIG_DEFAULT_DEVICE_TREE="qcom/sdm845-db845c"
++CONFIG_DEFAULT_DEVICE_TREE="qcom/sdm845-ayn-odin-m2"
+ CONFIG_SYS_LOAD_ADDR=0xA0000000
+ CONFIG_BUTTON_CMD=y
+ CONFIG_FIT=y
+ CONFIG_FIT_VERBOSE=y
+ CONFIG_BOOTSTD_FULL=y
++CONFIG_BOOTSTD_IGNORE_BOOTABLE=y
+ # CONFIG_BOOTMETH_VBE is not set
++# CONFIG_CMD_BOOTEFI_BOOTMGR is not set
+ CONFIG_BOOTDELAY=1
+ CONFIG_OF_BOARD_SETUP=y
+ CONFIG_USE_PREBOOT=y
+@@ -34,10 +36,13 @@ CONFIG_CMD_I2C=y
+ CONFIG_CMD_MMC=y
+ CONFIG_CMD_UFS=y
+ CONFIG_CMD_USB=y
++CONFIG_CMD_USB_MASS_STORAGE=y
+ CONFIG_CMD_CAT=y
+ CONFIG_CMD_BMP=y
+ CONFIG_CMD_EFIDEBUG=y
+ CONFIG_CMD_REGULATOR=y
++CONFIG_CMD_FAT=y
++CONFIG_CMD_FS_GENERIC=y
+ CONFIG_CMD_LOG=y
+ CONFIG_OF_LIVE=y
+ CONFIG_USE_DEFAULT_ENV_FILE=y
+@@ -58,6 +63,10 @@ CONFIG_CLK_QCOM_SC7280=y
+ CONFIG_DFU_MMC=y
+ CONFIG_DFU_SCSI=y
+ CONFIG_SYS_DFU_DATA_BUF_SIZE=0x200000
++CONFIG_USB_FUNCTION_FASTBOOT=y
++CONFIG_FASTBOOT_BUF_ADDR=0x1
++CONFIG_FASTBOOT_FLASH=y
++CONFIG_FASTBOOT_FLASH_MMC_DEV=0
+ CONFIG_MSM_GPIO=y
+ CONFIG_QCOM_PMIC_GPIO=y
+ CONFIG_DM_I2C=y
+@@ -114,11 +123,15 @@ CONFIG_SPMI_MSM=y
+ CONFIG_SYSINFO=y
+ CONFIG_SYSINFO_SMBIOS=y
+ CONFIG_USB=y
++CONFIG_DM_USB_GADGET=y
+ CONFIG_USB_XHCI_HCD=y
+ CONFIG_USB_XHCI_DWC3=y
+ CONFIG_USB_DWC3=y
+ CONFIG_USB_DWC3_GENERIC=y
++CONFIG_USB_STORAGE=y
+ CONFIG_USB_GADGET=y
++CONFIG_USB_GADGET_VENDOR_NUM=0x18d1
++CONFIG_USB_GADGET_PRODUCT_NUM=0xd00d
+ CONFIG_USB_GADGET_DOWNLOAD=y
+ CONFIG_USB_FUNCTION_MASS_STORAGE=y
+ CONFIG_UFS=y
+@@ -129,9 +142,11 @@ CONFIG_VIDEO_FONT_16X32=y
+ CONFIG_SYS_WHITE_ON_BLACK=y
+ CONFIG_NO_FB_CLEAR=y
+ CONFIG_VIDEO_SIMPLE=y
+-CONFIG_AUTOBOOT_PROMPT="Press the power button to enter menu\n"
+-CONFIG_AUTOBOOT_USE_MENUKEY=y
+-# \r (power button)
+-CONFIG_AUTOBOOT_MENUKEY=13
++CONFIG_SYS_CONSOLE_INFO_QUIET=y
++CONFIG_AUTOBOOT_KEYED=y
++CONFIG_AUTOBOOT_FLUSH_STDIN=y
++CONFIG_AUTOBOOT_PROMPT="Press power button to stop autoboot %d"
++CONFIG_AUTOBOOT_STOP_STR="\r"
+ CONFIG_BOOT_RETRY=y
+ CONFIG_BOOT_RETRY_TIME=2
++CONFIG_USB_FUNCTION_ACM=y
diff -rupbN u-boot.orig/arch/arm/dts/sdm845-ayn-odin-u-boot.dtsi u-boot/arch/arm/dts/sdm845-ayn-odin-u-boot.dtsi
--- u-boot.orig/arch/arm/dts/sdm845-ayn-odin-u-boot.dtsi	1970-01-01 00:00:00.000000000 +0000
+++ u-boot/arch/arm/dts/sdm845-ayn-odin-u-boot.dtsi	2025-07-08 16:42:33.729225375 +0000
@@ -0,0 +1,41 @@
+/ {
+	framebuffer: framebuffer@9d400000 {
+		compatible = "simple-framebuffer";
+		reg = <0 0x9d400000 0 0x2400000>;
+		width = <1080>;
+		height = <1920>;
+		stride = <(1080 * 4)>;
+		format = "a8r8g8b8";
+	};
+};
+
+&sdhc_2 {
+	/* Remove the unsupported rpmhcc xo clock reference */
+	clocks = <&gcc GCC_SDCC2_AHB_CLK>,
+		 <&gcc GCC_SDCC2_APPS_CLK>;
+	clock-names = "iface", "core";
+};
+
+&usb_1 {
+	status = "okay";
+	qcom,select-utmi-as-pipe-clk;
+};
+
+&usb_1_dwc3 {
+	dr_mode = "peripheral";
+
+	/* We don't support superspeed in u-boot so remove the SSPHY */
+	phys = <&usb_1_hsphy>;
+	phy-names = "usb2-phy";
+
+	/* fastest mode for USB 2 */
+	maximum-speed = "high-speed";
+};
+
+&usb_1_hsphy {
+	/* Remove ref clk because rmphcc isn't available */
+	clocks = <&gcc GCC_USB_PHY_CFG_AHB2PHY_CLK>;
+	clock-names = "cfg_ahb";
+
+	status = "okay";
+};
diff -rupbN u-boot.orig/board/qualcomm/default.env u-boot/board/qualcomm/default.env
--- u-boot.orig/board/qualcomm/default.env	2025-07-08 16:51:59.314597798 +0000
+++ u-boot/board/qualcomm/default.env	2025-07-08 17:27:02.813627295 +0000
@@ -3,9 +3,12 @@ stdout=serial,vidconsole
 stderr=serial,vidconsole
 preboot=scsi scan; usb start
 fastboot=fastboot -l $fastboot_addr_r usb 0
-do_boot=bootefi bootmgr
+do_boot=bootflow scan -lb
+serial_gadget=setenv stdin serial,button-kbd,usbacm; setenv stdout serial,vidconsole,usbacm; setenv stderr serial,vidconsole,usbacm; setenv bootretry -1; echo Enabled U-Boot console gadget :3
 bootmenu_0=Boot first available device=run do_boot
-bootmenu_1=Enable fastboot mode=run fastboot
-bootmenu_2=Reset device=reset
-menucmd=bootmenu
+bootmenu_1=Enable USB mass storage=ums 0 mmc 0
+bootmenu_2=Enable fastboot mode=run fastboot
+bootmenu_3=Enable serial console gadget=run serial_gadget
+bootmenu_4=Reset device=reset
+menucmd=bootmenu -1
 bootcmd=run do_boot
diff -rupbN u-boot.orig/common/cli_hush.c u-boot/common/cli_hush.c
--- u-boot.orig/common/cli_hush.c	2025-07-08 16:51:59.374599329 +0000
+++ u-boot/common/cli_hush.c	2025-07-08 17:29:14.181139021 +0000
@@ -1031,7 +1031,7 @@ static void get_user_input(struct in_str
 #  elif IS_ENABLED(CONFIG_RETRY_BOOTCMD)
 	strcpy(console_buffer, "run bootcmd\n");
 # else
-#	error "This only works with CONFIG_RESET_TO_RETRY or CONFIG_BOOT_RETRY_COMMAND enabled"
+	strcpy(console_buffer, "run menucmd\n");
 #  endif
 	}
 #endif
diff -rupbN u-boot.orig/configs/qcom_defconfig u-boot/configs/qcom_defconfig
--- u-boot.orig/configs/qcom_defconfig	2025-07-08 16:52:31.655422723 +0000
+++ u-boot/configs/qcom_defconfig	2025-07-08 17:34:14.896932130 +0000
@@ -4,13 +4,15 @@ CONFIG_POSITION_INDEPENDENT=y
 CONFIG_SYS_INIT_SP_BSS_OFFSET=1572864
 CONFIG_ARCH_SNAPDRAGON=y
 CONFIG_NR_DRAM_BANKS=24
-CONFIG_DEFAULT_DEVICE_TREE="qcom/sdm845-db845c"
+CONFIG_DEFAULT_DEVICE_TREE="qcom/sdm845-ayn-odin"
 CONFIG_SYS_LOAD_ADDR=0xA0000000
 CONFIG_BUTTON_CMD=y
 CONFIG_FIT=y
 CONFIG_FIT_VERBOSE=y
 CONFIG_BOOTSTD_FULL=y
+CONFIG_BOOTSTD_IGNORE_BOOTABLE=y
 # CONFIG_BOOTMETH_VBE is not set
+CONFIG_CMD_BOOTEFI_BOOTMGR is not set
 CONFIG_BOOTDELAY=1
 CONFIG_OF_BOARD_SETUP=y
 CONFIG_USE_PREBOOT=y
@@ -34,11 +36,14 @@ CONFIG_CMD_I2C=y
 CONFIG_CMD_MMC=y
 CONFIG_CMD_UFS=y
 CONFIG_CMD_USB=y
+CONFIG_CMD_USB_MASS_STORAGE=y
 CONFIG_CMD_CAT=y
 CONFIG_CMD_BMP=y
 CONFIG_CMD_EFIDEBUG=y
 CONFIG_CMD_RNG=y
 CONFIG_CMD_REGULATOR=y
+CONFIG_CMD_FAT=y
+CONFIG_CMD_FS_GENERIC=y
 CONFIG_CMD_LOG=y
 CONFIG_OF_LIVE=y
 CONFIG_OF_UPSTREAM_BUILD_VENDOR=y
@@ -64,7 +69,7 @@ CONFIG_DFU_MMC=y
 CONFIG_DFU_SCSI=y
 CONFIG_SYS_DFU_DATA_BUF_SIZE=0x200000
 CONFIG_USB_FUNCTION_FASTBOOT=y
-CONFIG_FASTBOOT_BUF_ADDR=0x0
+CONFIG_FASTBOOT_BUF_ADDR=0x1
 CONFIG_FASTBOOT_FLASH=y
 CONFIG_FASTBOOT_FLASH_MMC_DEV=0
 CONFIG_FASTBOOT_MMC_USER_SUPPORT=y
@@ -129,11 +134,15 @@ CONFIG_SYSINFO=y
 CONFIG_SYSINFO_SMBIOS=y
 CONFIG_SYSRESET_QCOM_PSHOLD=y
 CONFIG_USB=y
+CONFIG_DM_USB_GADGET=y
 CONFIG_USB_XHCI_HCD=y
 CONFIG_USB_XHCI_DWC3=y
 CONFIG_USB_DWC3=y
 CONFIG_USB_DWC3_GENERIC=y
+CONFIG_USB_STORAGE=y
 CONFIG_USB_GADGET=y
+CONFIG_USB_GADGET_VENDOR_NUM=0x18d1
+CONFIG_USB_GADGET_PRODUCT_NUM=0xd00d
 CONFIG_USB_FUNCTION_MASS_STORAGE=y
 CONFIG_UFS=y
 CONFIG_QCOM_UFS=y
@@ -143,3 +152,11 @@ CONFIG_VIDEO_FONT_16X32=y
 CONFIG_SYS_WHITE_ON_BLACK=y
 CONFIG_NO_FB_CLEAR=y
 CONFIG_VIDEO_SIMPLE=y
+CONFIG_SYS_CONSOLE_INFO_QUIET=y
+CONFIG_AUTOBOOT_KEYED=y
+CONFIG_AUTOBOOT_FLUSH_STDIN=y
+CONFIG_AUTOBOOT_PROMPT="Press power button to stop autoboot %d"
+CONFIG_AUTOBOOT_STOP_STR="\r"
+CONFIG_BOOT_RETRY=y
+CONFIG_BOOT_RETRY_TIME=2
+CONFIG_USB_FUNCTION_ACM=y
diff -rupbN u-boot.orig/dts/upstream/src/arm64/qcom/sdm845-ayn-odin.dts u-boot/dts/upstream/src/arm64/qcom/sdm845-ayn-odin.dts
--- u-boot.orig/dts/upstream/src/arm64/qcom/sdm845-ayn-odin.dts	1970-01-01 00:00:00.000000000 +0000
+++ u-boot/dts/upstream/src/arm64/qcom/sdm845-ayn-odin.dts	2025-07-08 16:40:36.818456882 +0000
@@ -0,0 +1,1057 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2022, Teguh Sobirin.
+ * Copyright (c) 2024, Anton Bambura.
+ */
+
+/dts-v1/;
+
+#include <dt-bindings/pinctrl/qcom,pmic-gpio.h>
+#include <dt-bindings/regulator/qcom,rpmh-regulator.h>
+#include <dt-bindings/sound/qcom,q6afe.h>
+#include <dt-bindings/sound/qcom,q6asm.h>
+
+#include "sdm845.dtsi"
+#include "sdm845-wcd9340.dtsi"
+#include "pm8998.dtsi"
+#include "pmi8998.dtsi"
+
+/ {
+	model = "AYN Odin";
+	compatible = "ayn,odin", "qcom,sdm845";
+	qcom,msm-id = <321 0x20001>;
+	qcom,board-id = <8 0>;
+	chassis-type = "tablet";
+
+	aliases {
+		serial0 = &uart9;
+		hsuart0 = &uart6;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	/* Fixed crystal oscillator dedicated to MCP2517FD */
+	clk40M: can-clock {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-frequency = <40000000>;
+	};
+
+	gpio-keys {
+		compatible = "gpio-keys";
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&volume_up_gpio>;
+
+		key-volume-up {
+			label = "Volume Up";
+			linux,code = <KEY_VOLUMEUP>;
+			gpios = <&pm8998_gpios 6 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	odin-gamepad {
+		compatible = "odin-gamepad";
+
+		gamepad-name = "Odin Gamepad";
+		gamepad-vendor = <0x2020>;
+		gamepad-product = <0x0111>;
+		gamepad-revision = <0x0001>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&btn_pins &btn_en_active>;
+
+		io-channel-names = "abs_x", "abs_y",
+				 "abs_rx", "abs_ry",
+				 "abs_z", "abs_rz";
+		io-channels = <&pm8998_adc ADC5_GPIO2>,
+					  <&pm8998_adc ADC5_GPIO3>,
+					  <&pm8998_adc ADC5_GPIO5>,
+					  <&pm8998_adc ADC5_GPIO6_DIV3>,
+					  <&pm8998_adc ADC5_GPIO1>,
+					  <&pm8998_adc ADC5_GPIO4>;
+
+		poll-interval = <10>;
+
+		adc-count = <6>;
+		adc-fuzz = <30>;
+		adc-flat = <30>;
+		adc-deadzone = <200>;
+		adc-x-range = <12000>;
+		adc-y-range = <12000>;
+		adc-rx-range = <12000>;
+		adc-ry-range = <12000>;
+		adc-z-range = <12000>;
+		adc-rz-range = <12000>;
+
+		btn_gpio01 {
+			gpios = <&tlmm 96 GPIO_ACTIVE_LOW>;
+			label = "btn_gpio01";
+			linux,code = <BTN_EAST>;
+		};
+
+		btn_gpio02 {
+			gpios = <&tlmm 95 GPIO_ACTIVE_LOW>;
+			label = "btn_gpio02";
+			linux,code = <BTN_SOUTH>;
+		};
+
+		btn_gpio03 {
+			gpios = <&tlmm 41 GPIO_ACTIVE_LOW>;
+			label = "btn_gpio03";
+			linux,code = <BTN_WEST>;
+		};
+
+		btn_gpio04 {
+			gpios = <&tlmm 92 GPIO_ACTIVE_LOW>;
+			label = "btn_gpio04";
+			linux,code = <BTN_NORTH>;
+		};
+
+		btn_gpio05 {
+			gpios = <&tlmm 26 GPIO_ACTIVE_LOW>;
+			label = "btn_gpio05";
+			linux,code = <BTN_TL>;
+		};
+
+		btn_gpio06 {
+			gpios = <&tlmm 91 GPIO_ACTIVE_LOW>;
+			label = "btn_gpio06";
+			linux,code = <BTN_TR>;
+		};
+
+		btn_gpio07 {
+			gpios = <&tlmm 0 GPIO_ACTIVE_LOW>;
+			label = "btn_gpio07";
+			linux,code = <BTN_SELECT>;
+		};
+
+		btn_gpio08 {
+			gpios = <&tlmm 3 GPIO_ACTIVE_LOW>;
+			label = "btn_gpio08";
+			linux,code = <BTN_START>;
+		};
+
+		btn_gpio09 {
+			gpios = <&tlmm 40 GPIO_ACTIVE_LOW>;
+			label = "btn_gpio09";
+			linux,code = <BTN_THUMBL>;
+		};
+
+		btn_gpio10 {
+			gpios = <&tlmm 22 GPIO_ACTIVE_LOW>;
+			label = "btn_gpio10";
+			linux,code = <BTN_THUMBR>;
+		};
+
+		btn_gpio11 {
+			gpios = <&tlmm 124 GPIO_ACTIVE_LOW>;
+			label = "btn_gpio11";
+			linux,code = <BTN_DPAD_UP>;
+		};
+
+		btn_gpio12 {
+			gpios = <&tlmm 128 GPIO_ACTIVE_LOW>;
+			label = "btn_gpio12";
+			linux,code = <BTN_DPAD_DOWN>;
+		};
+
+		btn_gpio13 {
+			gpios = <&tlmm 52 GPIO_ACTIVE_LOW>;
+			label = "btn_gpio13";
+			linux,code = <BTN_DPAD_LEFT>;
+		};
+
+		btn_gpio14 {
+			gpios = <&tlmm 122 GPIO_ACTIVE_LOW>;
+			label = "btn_gpio14";
+			linux,code = <BTN_DPAD_RIGHT>;
+		};
+
+		btn_gpio15 {
+			gpios = <&tlmm 49 GPIO_ACTIVE_LOW>;
+			label = "btn_gpio15";
+			linux,code = <BTN_MODE>;
+		};
+	};
+
+	battery: battery {
+		compatible = "simple-battery";
+
+		charge-full-design-microamp-hours = <3850000>;
+		voltage-min-design-microvolt = <3600000>;
+		voltage-max-design-microvolt = <4400000>;
+	};
+
+	connector {
+		compatible = "hdmi-connector";
+
+		label = "hdmi";
+		type = "d";
+
+		hdmi-pwr-supply = <&hdmi_en_reg>;
+		hpd-gpios = <&tlmm 121 GPIO_ACTIVE_HIGH>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&hdmi_hpd>;
+
+		port {
+			hdmi_con: endpoint {
+				remote-endpoint = <&lt8912b_out>;
+			};
+		};
+	};
+
+	hdmi_en_reg: hdmi-en-regulator {
+		compatible = "regulator-fixed";
+		regulator-name = "hdmi_en";
+
+		gpio = <&tlmm 13 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+
+	vph_pwr: vph-pwr-regulator {
+		compatible = "regulator-fixed";
+		regulator-name = "vph_pwr";
+		regulator-min-microvolt = <3700000>;
+		regulator-max-microvolt = <3700000>;
+	};
+
+	vreg_s4a_1p8: pm8998-smps4 {
+		compatible = "regulator-fixed";
+		regulator-name = "vreg_s4a_1p8";
+
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+
+		regulator-always-on;
+		regulator-boot-on;
+
+		vin-supply = <&vph_pwr>;
+	};
+};
+
+&adsp_pas {
+	status = "okay";
+	firmware-name = "qcom/sdm845/adsp.mbn";
+};
+
+&apps_rsc {
+	pm8998-rpmh-regulators {
+		compatible = "qcom,pm8998-rpmh-regulators";
+		qcom,pmic-id = "a";
+
+		vdd-s1-supply = <&vph_pwr>;
+		vdd-s2-supply = <&vph_pwr>;
+		vdd-s3-supply = <&vph_pwr>;
+		vdd-s4-supply = <&vph_pwr>;
+		vdd-s5-supply = <&vph_pwr>;
+		vdd-s6-supply = <&vph_pwr>;
+		vdd-s7-supply = <&vph_pwr>;
+		vdd-s8-supply = <&vph_pwr>;
+		vdd-s9-supply = <&vph_pwr>;
+		vdd-s10-supply = <&vph_pwr>;
+		vdd-s11-supply = <&vph_pwr>;
+		vdd-s12-supply = <&vph_pwr>;
+		vdd-s13-supply = <&vph_pwr>;
+		vdd-l1-l27-supply = <&vreg_s7a_1p025>;
+		vdd-l2-l8-l17-supply = <&vreg_s3a_1p35>;
+		vdd-l3-l11-supply = <&vreg_s7a_1p025>;
+		vdd-l4-l5-supply = <&vreg_s7a_1p025>;
+		vdd-l6-supply = <&vph_pwr>;
+		vdd-l7-l12-l14-l15-supply = <&vreg_s5a_2p04>;
+		vdd-l9-supply = <&vreg_bob>;
+		vdd-l10-l23-l25-supply = <&vreg_bob>;
+		vdd-l13-l19-l21-supply = <&vreg_bob>;
+		vdd-l16-l28-supply = <&vreg_bob>;
+		vdd-l18-l22-supply = <&vreg_bob>;
+		vdd-l20-l24-supply = <&vreg_bob>;
+		vdd-l26-supply = <&vreg_s3a_1p35>;
+		vin-lvs-1-2-supply = <&vreg_s4a_1p8>;
+
+		vreg_s3a_1p35: smps3 {
+			regulator-min-microvolt = <1352000>;
+			regulator-max-microvolt = <1352000>;
+		};
+
+		vreg_s5a_2p04: smps5 {
+			regulator-min-microvolt = <1904000>;
+			regulator-max-microvolt = <2040000>;
+		};
+
+		vreg_s7a_1p025: smps7 {
+			regulator-min-microvolt = <900000>;
+			regulator-max-microvolt = <1028000>;
+		};
+
+		vdda_mipi_dsi0_pll:
+		vdda_mipi_dsi1_pll:
+		vdda_qlink_lv:
+		vdda_ufs1_core:
+		vdda_usb1_ss_core:
+		vdda_qusb_hs0_1p8:
+		vdda_qusb_hs0_3p1:
+		vreg_l1a_0p875: ldo1 {
+			regulator-min-microvolt = <880000>;
+			regulator-max-microvolt = <880000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l5a_0p8: ldo5 {
+			regulator-min-microvolt = <800000>;
+			regulator-max-microvolt = <800000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l7a_1p8: ldo7 {
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l12a_1p8: ldo12 {
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l13a_2p95: ldo13 {
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <2960000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l14a_1p8: ldo14 {
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+			regulator-boot-on;
+			regulator-always-on;
+		};
+
+		vreg_l17a_1p3: ldo17 {
+			regulator-min-microvolt = <1304000>;
+			regulator-max-microvolt = <1304000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l20a_2p95: ldo20 {
+			regulator-min-microvolt = <2950000>;
+			regulator-max-microvolt = <2960000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l21a_2p95: ldo21 {
+			regulator-min-microvolt = <2950000>;
+			regulator-max-microvolt = <2960000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l23a_3p3: ldo23 {
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3312000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l24a_3p075: ldo24 {
+			regulator-min-microvolt = <3088000>;
+			regulator-max-microvolt = <3088000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l25a_3p3: ldo25 {
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3312000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l26a_1p2: ldo26 {
+			regulator-min-microvolt = <1200000>;
+			regulator-max-microvolt = <1200000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+			regulator-boot-on;
+		};
+
+		vreg_l28a_3p0: ldo28 {
+			regulator-min-microvolt = <2856000>;
+			regulator-max-microvolt = <3008000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+	};
+
+	pmi8998-rpmh-regulators {
+		compatible = "qcom,pmi8998-rpmh-regulators";
+		qcom,pmic-id = "b";
+
+		vdd-bob-supply = <&vph_pwr>;
+
+		vreg_bob: bob {
+			regulator-min-microvolt = <3312000>;
+			regulator-max-microvolt = <3600000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_AUTO>;
+			regulator-allow-bypass;
+		};
+	};
+
+	pm8005-rpmh-regulators {
+		compatible = "qcom,pm8005-rpmh-regulators";
+		qcom,pmic-id = "c";
+
+		vdd-s1-supply = <&vph_pwr>;
+		vdd-s2-supply = <&vph_pwr>;
+		vdd-s3-supply = <&vph_pwr>;
+		vdd-s4-supply = <&vph_pwr>;
+
+		vreg_s3c_0p6: smps3 {
+			regulator-min-microvolt = <600000>;
+			regulator-max-microvolt = <600000>;
+		};
+	};
+};
+
+&cdsp_pas {
+	status = "okay";
+	firmware-name = "qcom/sdm845/cdsp.mbn";
+};
+
+&mdss {
+	status = "okay";
+};
+
+&mdss_dsi0 {
+	status = "okay";
+	vdda-supply = <&vreg_l26a_1p2>;
+
+	display_panel: panel@0 {
+		compatible = "innolux,td4328";
+		status = "okay";
+		reg = <0>;
+		vddio-supply = <&vreg_l14a_1p8>;
+		vddpos-supply = <&lab>;
+		vddneg-supply = <&ibb>;
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		backlight = <&pmi8998_wled>;
+		reset-gpios = <&tlmm 6 GPIO_ACTIVE_LOW>;
+		rotation = <90>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&panel_reset_active>;
+
+		port {
+			panel_in: endpoint {
+				remote-endpoint = <&mdss_dsi0_out>;
+			};
+		};
+	};
+};
+
+&mdss_dsi0_out {
+	remote-endpoint = <&panel_in>;
+	data-lanes = <0 1 2 3>;
+};
+
+&mdss_dsi0_phy {
+	status = "okay";
+	vdds-supply = <&vdda_mipi_dsi0_pll>;
+};
+
+&mdss_dsi1 {
+	vdda-supply = <&vreg_l26a_1p2>;
+	status = "okay";
+};
+
+&mdss_dsi1_out {
+	remote-endpoint = <&lt8912b_in>;
+	data-lanes = <0 1 2 3>;
+};
+
+&mdss_dsi1_phy {
+	vdds-supply = <&vreg_l1a_0p875>;
+	status = "okay";
+};
+
+&dispcc {
+	status = "okay";
+};
+
+&gcc {
+	protected-clocks = <GCC_QSPI_CORE_CLK>,
+			   <GCC_QSPI_CORE_CLK_SRC>,
+			   <GCC_QSPI_CNOC_PERIPH_AHB_CLK>,
+			   <GCC_LPASS_Q6_AXI_CLK>,
+			   <GCC_LPASS_SWAY_CLK>;
+};
+
+&gmu {
+	status = "okay";
+};
+
+&gpi_dma0 {
+	status = "okay";
+};
+
+&gpi_dma1 {
+	status = "okay";
+};
+
+&gpu {
+	status = "okay";
+	zap-shader {
+		memory-region = <&gpu_mem>;
+		firmware-name = "qcom/sdm845/a630_zap.mbn";
+	};
+};
+
+&i2c4 {
+	status = "okay";
+	clock-frequency = <400000>;
+
+	hdmi-bridge@48 {
+		compatible = "lontium,lt8912b";
+		reg = <0x48>;
+		reset-gpios = <&tlmm 12 GPIO_ACTIVE_LOW>;
+
+		ports {
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			port@0 {
+				reg = <0>;
+
+				lt8912b_in: endpoint {
+					data-lanes = <0 1 2 3>;
+					remote-endpoint = <&mdss_dsi1_out>;
+				};
+			};
+
+			port@1 {
+				reg = <1>;
+
+				lt8912b_out: endpoint {
+					remote-endpoint = <&hdmi_con>;
+				};
+			};
+		};
+	};
+};
+
+&i2c5 {
+	status = "okay";
+	clock-frequency = <400000>;
+
+	synaptics-rmi4-i2c@20 {
+		compatible = "syna,rmi4-i2c";
+		reg = <0x20>;
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		reset-gpios = <&tlmm 104 GPIO_ACTIVE_LOW>;
+		interrupts-extended = <&tlmm 125 IRQ_TYPE_EDGE_FALLING>;
+
+		vdd-supply = <&vreg_l28a_3p0>;
+		vio-supply = <&vreg_l14a_1p8>;
+
+		syna,reset-delay-ms = <200>;
+		syna,startup-delay-ms = <200>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&ts_int_active &ts_reset_active>;
+
+		rmi4-f01@1 {
+			reg = <0x01>;
+			syna,nosleep-mode = <1>;
+		};
+
+		rmi4_f12: rmi4-f12@12 {
+			reg = <0x12>;
+			touchscreen-inverted-x;
+			touchscreen-x-mm = <75>;
+			touchscreen-y-mm = <132>;
+			syna,sensor-type = <1>;
+			syna,rezero-wait-ms = <200>;
+		};
+	};
+};
+
+&spi0 {
+	/*
+	 * TODO
+	 * [   16.605399] sdm845-pinctrl 3400000.pinctrl: pin GPIO_0 already requested by odin-gamepad; cannot claim for 880000.spi
+	 */
+	status = "disabled";
+	cs-gpios = <&tlmm 3 GPIO_ACTIVE_LOW>;
+
+	can@0 {
+		compatible = "microchip,mcp2517fd";
+		reg = <0>;
+		clocks = <&clk40M>;
+		interrupts-extended = <&tlmm 34 IRQ_TYPE_LEVEL_LOW>;
+		spi-max-frequency = <10000000>;
+	};
+};
+
+&ibb {
+	regulator-min-microvolt = <4600000>;
+	regulator-max-microvolt = <6000000>;
+	regulator-over-current-protection;
+	regulator-pull-down;
+	regulator-soft-start;
+	qcom,discharge-resistor-kohms = <300>;
+};
+
+&lab {
+	regulator-min-microvolt = <4600000>;
+	regulator-max-microvolt = <6000000>;
+	regulator-over-current-protection;
+	regulator-pull-down;
+	regulator-soft-start;
+};
+
+&mss_pil {
+	status = "okay";
+	firmware-name = "qcom/sdm845/mba.mbn", "qcom/sdm845/modem.mbn";
+};
+
+&pm8998_adc {
+	adc-chan@12 {
+		reg = <ADC5_GPIO1>;
+		label = "gpio1";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 1>;
+	};
+
+	adc-chan@13 {
+		reg = <ADC5_GPIO2>;
+		label = "gpio2";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 1>;
+	};
+
+	adc-chan@14 {
+		reg = <ADC5_GPIO3>;
+		label = "gpio3";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 1>;
+	};
+
+	adc-chan@15 {
+		reg = <ADC5_GPIO4>;
+		label = "gpio4";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 1>;
+	};
+
+	adc-chan@16 {
+		reg = <ADC5_GPIO5>;
+		label = "gpio5";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 1>;
+	};
+
+	adc-chan@97 {
+		reg = <ADC5_GPIO6_DIV3>;
+		label = "gpio6";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 3>;
+	};
+};
+
+&pm8998_gpios {
+	volume_up_gpio: vol-up-active {
+		pins = "gpio6";
+		function = "normal";
+		input-enable;
+		bias-pull-up;
+		qcom,drive-strength = <PMIC_GPIO_STRENGTH_NO>;
+	};
+};
+
+&pm8998_resin {
+	linux,code = <KEY_VOLUMEDOWN>;
+	status = "okay";
+};
+
+&pmi8998_charger {
+	status = "okay";
+	monitored-battery = <&battery>;
+};
+
+&pmi8998_rradc {
+	status = "okay";
+	monitored-battery = <&battery>;
+};
+
+&pmi8998_wled {
+	status = "okay";
+	qcom,enabled-strings = <0 1>;
+	qcom,num-strings = <2>;
+};
+
+&q6asmdai {
+	dai@0 {
+		reg = <0>;
+	};
+
+	dai@1 {
+		reg = <1>;
+	};
+
+	dai@2 {
+		reg = <2>;
+	};
+};
+
+&qupv3_id_0 {
+	status = "okay";
+};
+
+&qupv3_id_1 {
+	status = "okay";
+};
+
+&qup_uart6_default {
+	pinmux {
+		pins = "gpio45", "gpio46", "gpio47", "gpio48";
+		function = "qup6";
+	};
+
+	cts {
+		pins = "gpio45";
+		bias-disable;
+	};
+
+	rts-tx {
+		pins = "gpio46", "gpio47";
+		drive-strength = <2>;
+		bias-disable;
+	};
+
+	rx {
+		pins = "gpio48";
+		bias-pull-up;
+	};
+};
+
+&qup_uart9_default {
+	pinconf-tx {
+		pins = "gpio4";
+		drive-strength = <2>;
+		bias-disable;
+	};
+
+	pinconf-rx {
+		pins = "gpio5";
+		drive-strength = <2>;
+		bias-pull-up;
+	};
+};
+
+&sdhc_2 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&sdc2_default_state &sdc2_card_det_n>;
+
+	vmmc-supply = <&vreg_l21a_2p95>;
+	vqmmc-supply = <&vreg_l13a_2p95>;
+
+	bus-width = <4>;
+	cd-gpios = <&tlmm 126 GPIO_ACTIVE_LOW>;
+};
+
+&sound {
+	compatible = "ayn,odin-sndcard", "qcom,sdm845-sndcard";
+	model = "AYN Odin";
+
+	audio-routing =
+		"RX_BIAS", "MCLK",
+		"AMIC2", "MIC BIAS2",
+		"SpkrLeft IN", "SPK1 OUT",
+		"SpkrRight IN", "SPK2 OUT",
+		"MM_DL1",  "MultiMedia1 Playback",
+		"MM_DL3",  "MultiMedia3 Playback",
+		"MultiMedia2 Capture", "MM_UL2";
+
+	mm1-dai-link {
+		link-name = "MultiMedia1";
+		cpu {
+			sound-dai = <&q6asmdai  MSM_FRONTEND_DAI_MULTIMEDIA1>;
+		};
+	};
+
+	mm2-dai-link {
+		link-name = "MultiMedia2";
+		cpu {
+			sound-dai = <&q6asmdai  MSM_FRONTEND_DAI_MULTIMEDIA2>;
+		};
+	};
+
+	mm3-dai-link {
+		link-name = "MultiMedia3";
+		cpu {
+			sound-dai = <&q6asmdai  MSM_FRONTEND_DAI_MULTIMEDIA3>;
+		};
+	};
+
+	slim-dai-link {
+		link-name = "SLIM Playback";
+		cpu {
+			sound-dai = <&q6afedai SLIMBUS_0_RX>;
+		};
+
+		platform {
+			sound-dai = <&q6routing>;
+		};
+
+		codec {
+			sound-dai =  <&left_spkr>, <&right_spkr>, <&swm 0>, <&wcd9340 0>;
+		};
+	};
+
+	slimcap-dai-link {
+		link-name = "SLIM Capture";
+		cpu {
+			sound-dai = <&q6afedai SLIMBUS_0_TX>;
+		};
+
+		platform {
+			sound-dai = <&q6routing>;
+		};
+
+		codec {
+			sound-dai = <&wcd9340 1>;
+		};
+	};
+
+	slim-wcd-dai-link {
+		link-name = "SLIM WCD Playback";
+		cpu {
+			sound-dai = <&q6afedai SLIMBUS_1_RX>;
+		};
+
+		platform {
+			sound-dai = <&q6routing>;
+		};
+
+		codec {
+			sound-dai =  <&wcd9340 2>;
+		};
+	};
+};
+
+&slpi_pas {
+	firmware-name = "qcom/sdm845/AYN/Odin/slpi.mbn";
+	status = "okay";
+};
+
+&tlmm {
+	sdc2_default_state: sdc2-default {
+		clk {
+			pins = "sdc2_clk";
+			bias-disable;
+			drive-strength = <16>;
+		};
+
+		cmd {
+			pins = "sdc2_cmd";
+			bias-pull-up;
+			drive-strength = <10>;
+		};
+
+		data {
+			pins = "sdc2_data";
+			bias-pull-up;
+			drive-strength = <10>;
+		};
+	};
+
+	sdc2_card_det_n: sd-card-det-n {
+		pins = "gpio126";
+		function = "gpio";
+		bias-pull-up;
+	};
+
+	ts_int_active: ts-int-active {
+		pins = "gpio125";
+		function = "gpio";
+		drive-strength = <16>;
+		bias-pull-up;
+		input-enable;
+	};
+
+	ts_reset_active: ts-reset-active {
+		pins = "gpio104";
+		function = "gpio";
+		drive-strength = <16>;
+		bias-pull-up;
+	};
+
+	btn_en_active: btn-en-active {
+		pins = "gpio88";
+		function = "gpio";
+		drive-strength = <8>;
+		bias-disable;
+		output-high;
+	};
+
+	btn_pins: btn-pins {
+		mux {
+			pins =	"gpio96", "gpio95", "gpio41", "gpio92",
+				"gpio26", "gpio91", "gpio0", "gpio3",
+				"gpio40", "gpio22", "gpio124", "gpio128",
+				"gpio52", "gpio122", "gpio49";
+			function = "gpio";
+			drive-strength = <16>;
+			bias-pull-up;
+		};
+	};
+
+	panel_reset_active: panel-reset-active {
+		pins = "gpio6";
+		function = "gpio";
+		drive-strength = <16>;
+		bias-disable;
+	};
+
+	hdmi_hpd: hdmi-hpd-active {
+		pins = "gpio121";
+		function = "gpio";
+		drive-strength = <10>;
+		bias-pull-up;
+		input-enable;
+	};
+};
+
+&uart6 {
+	status = "okay";
+
+	bluetooth {
+		compatible = "qcom,wcn3990-bt";
+
+		vddio-supply = <&vreg_s4a_1p8>;
+		vddxo-supply = <&vreg_l7a_1p8>;
+		vddrf-supply = <&vreg_l17a_1p3>;
+		vddch0-supply = <&vreg_l25a_3p3>;
+		vddch1-supply = <&vreg_l23a_3p3>;
+		max-speed = <3200000>;
+	};
+};
+
+&uart9 {
+	status = "okay";
+};
+
+&ufs_mem_hc {
+	status = "okay";
+
+	reset-gpios = <&tlmm 150 GPIO_ACTIVE_LOW>;
+
+	vcc-supply = <&vreg_l20a_2p95>;
+	vcc-max-microamp = <600000>;
+};
+
+&ufs_mem_phy {
+	status = "okay";
+
+	vdda-phy-supply = <&vreg_l1a_0p875>;
+	vdda-pll-supply = <&vreg_l26a_1p2>;
+};
+
+&usb_1 {
+	status = "okay";
+};
+
+&usb_1_dwc3 {
+	dr_mode = "peripheral";
+};
+
+&usb_1_hsphy {
+	status = "okay";
+
+	vdd-supply = <&vreg_l1a_0p875>;
+	vdda-pll-supply = <&vreg_l12a_1p8>;
+	vdda-phy-dpdm-supply = <&vreg_l24a_3p075>;
+
+	qcom,imp-res-offset-value = <8>;
+	qcom,hstx-trim-value = <QUSB2_V2_HSTX_TRIM_21_6_MA>;
+	qcom,preemphasis-level = <QUSB2_V2_PREEMPHASIS_5_PERCENT>;
+	qcom,preemphasis-width = <QUSB2_V2_PREEMPHASIS_WIDTH_HALF_BIT>;
+};
+
+&usb_1_qmpphy {
+	status = "okay";
+
+	vdda-phy-supply = <&vreg_l26a_1p2>;
+	vdda-pll-supply = <&vreg_l1a_0p875>;
+};
+
+&venus {
+	status = "okay";
+	firmware-name = "qcom/venus-5.2/venus.mbn";
+};
+
+&wcd9340 {
+	reset-gpios = <&tlmm 64 GPIO_ACTIVE_HIGH>;
+	vdd-buck-supply = <&vreg_s4a_1p8>;
+	vdd-buck-sido-supply = <&vreg_s4a_1p8>;
+	vdd-tx-supply = <&vreg_s4a_1p8>;
+	vdd-rx-supply = <&vreg_s4a_1p8>;
+	vdd-io-supply = <&vreg_s4a_1p8>;
+	qcom,mbhc-buttons-vthreshold-microvolt = <75000 150000 237000 500000 500000 500000 500000 500000>;
+	qcom,mbhc-headset-vthreshold-microvolt = <1700000>;
+	qcom,mbhc-headphone-vthreshold-microvolt = <50000>;
+};
+
+&swm {
+	left_spkr: wsa8810-left{
+		compatible = "sdw10217201000";
+		reg = <0 1>;
+		powerdown-gpios = <&wcdgpio 2 GPIO_ACTIVE_HIGH>;
+		#thermal-sensor-cells = <0>;
+		sound-name-prefix = "SpkrLeft";
+		#sound-dai-cells = <0>;
+	};
+
+	right_spkr: wsa8810-right{
+		compatible = "sdw10217201000";
+		powerdown-gpios = <&wcdgpio 2 GPIO_ACTIVE_HIGH>;
+		reg = <0 2>;
+		#thermal-sensor-cells = <0>;
+		sound-name-prefix = "SpkrRight";
+		#sound-dai-cells = <0>;
+	};
+};
+
+&wifi {
+	status = "okay";
+
+	vdd-0.8-cx-mx-supply = <&vreg_l5a_0p8>;
+	vdd-1.8-xo-supply = <&vreg_l7a_1p8>;
+	vdd-1.3-rfa-supply = <&vreg_l17a_1p3>;
+	vdd-3.3-ch0-supply = <&vreg_l25a_3p3>;
+	vdd-3.3-ch1-supply = <&vreg_l23a_3p3>;
+
+	qcom,snoc-host-cap-8bit-quirk;
+};
+
+/* PINCTRL - additions to nodes defined in sdm845.dtsi */
+&qup_spi0_default {
+	drive-strength = <6>;
+	bias-disable;
+};
+
+&qup_i2c4_default {
+	drive-strength = <2>;
+	bias-disable;
+};
+
+&qup_i2c5_default {
+	drive-strength = <2>;
+	bias-disable;
+};
diff -rupbN u-boot.orig/include/common.h u-boot/include/common.h
--- u-boot.orig/include/common.h	1970-01-01 00:00:00.000000000 +0000
+++ u-boot/include/common.h	2025-07-08 16:39:12.972468136 +0000
@@ -0,0 +1,33 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * Common header file for U-Boot
+ *
+ * This file still includes quite a few headers that should be included
+ * individually as needed. Patches to remove things are welcome.
+ *
+ * (C) Copyright 2000-2009
+ * Wolfgang Denk, DENX Software Engineering, wd@denx.de.
+ */
+
+#ifndef __COMMON_H_
+#define __COMMON_H_	1
+
+#ifndef __ASSEMBLY__		/* put C only stuff in this section */
+#include <config.h>
+#include <errno.h>
+#include <time.h>
+#include <linux/types.h>
+#include <linux/string.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <linux/kernel.h>
+#include <asm/u-boot.h> /* boot information for Linux kernel */
+#include <vsprintf.h>
+#endif	/* __ASSEMBLY__ */
+
+/* Pull in stuff for the build system */
+#ifdef DO_DEPS_ONLY
+# include <env_internal.h>
+#endif
+
+#endif	/* __COMMON_H_ */
